slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/BlockImporter.java

```diff
@@ -0,0 +1,122 @@
+  private final LowWatermarkDao lowWatermarkDao;
+  private final SignedBlocksDao signedBlocksDao;
+  private final Validator validator;
+  private final Handle handle;
+  private final ObjectMapper mapper;
+
+  public BlockImporter(
+      final Validator validator,
+      final Handle handle,
+      final ObjectMapper mapper,
+      final LowWatermarkDao lowWatermarkDao,
+      final SignedBlocksDao signedBlocksDao) {
+    this.validator = validator;
+    this.handle = handle;
+    this.mapper = mapper;
+    this.lowWatermarkDao = lowWatermarkDao;
+    this.signedBlocksDao = signedBlocksDao;
+  }
+
+  public void importFrom(final ArrayNode signedBlocksNode) throws JsonProcessingException {
+
+    for (int i = 0; i < signedBlocksNode.size(); i++) {
+      final SignedBlock jsonBlock = mapper.treeToValue(signedBlocksNode.get(i), SignedBlock.class);
+      final BlockValidator blockValidator =
+          new BlockValidator(
+              handle,
+              jsonBlock.getSigningRoot(),
+              jsonBlock.getSlot(),
+              validator.getId(),
+              signedBlocksDao,
+              lowWatermarkDao);
+
+      final String blockIdentifierString =
+          String.format("Block with index %d for validator %s", i, validator.getPublicKey());
+
+      if (jsonBlock.getSigningRoot() == null) {
+        if (nullBlockAlreadyExistsInSlot(jsonBlock.getSlot())) {
+          LOG.warn("{} - already exists in database, not imported", blockIdentifierString);
+        } else {
+          persist(jsonBlock);
+        }
+      } else {
+        if (blockValidator.directlyConflictsWithExistingEntry()) {
+          LOG.warn("{} - conflicts with an existing entry", blockIdentifierString);
+        }
+
+        if (blockValidator.alreadyExists()) {
+          LOG.debug("{} - already exists in database, not imported", blockIdentifierString);
+        } else {
+          persist(jsonBlock);
+        }
+      }
+    }
+
+    final Optional<SigningWatermark> watermark =
+        lowWatermarkDao.findLowWatermarkForValidator(handle, validator.getId());
+
+    if (minSlotTracker.compareTrackedValueTo(watermark.map(SigningWatermark::getSlot)) > 0) {
+      LOG.warn(
+          "Updating Block slot low watermark to {}", minSlotTracker.getTrackedMinValue().get());
+      lowWatermarkDao.updateSlotWatermarkFor(
+          handle, validator.getId(), minSlotTracker.getTrackedMinValue().get());
+    }
+  }
+
+  private void persist(final SignedBlock jsonBlock) {
+    signedBlocksDao.insertBlockProposal(
+        handle,
+        new tech.pegasys.web3signer.slashingprotection.dao.SignedBlock(
+            validator.getId(), jsonBlock.getSlot(), jsonBlock.getSigningRoot()));
+    minSlotTracker.trackValue(jsonBlock.getSlot());
+  }
+
+  private boolean nullBlockAlreadyExistsInSlot(final UInt64 slot) {
+    return handle
+        .createQuery(
+            "SELECT validator_id, slot, signing_root "
```

Comment:
same comment as before, can be replaced with count(*) instead of bringing all the records back.

Reply:
done